Support (m)TLS API Socket

meln5674 opened the pull request:
Add flags `--tls-cert`, `--tls-key`, `--tls-ca`/`--tls-client-ca` to the commands `podman remote`, `podman system service`, and `podman system connection add` to support serving the API socket using TLS and mTLS, as well as connecting to such a socket.

This relies on https://github.com/containers/common/pull/2249 and will fail CI until merged.

Fixes https://github.com/containers/podman/issues/24583

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
* The `podman system service` command now supports serving over tcp with TLS and mTLS
* The `podman system connection add` command now supports creating connections to TLS and mTLS tcp sockets
* The `podman remote` commands now support connecting to TLS and mTLS tcp sockets
```

Luap99 commented:
@jwhonce @baude @mheon FYI. Had you ever looked at TLS support for the podman service and remote client?
With the amount of code here this looks easy enough to accept to me but I was wondering if there is/was any reason why we do not support it.

mheon commented:
Just skimmed, but LGTM on approach. A lot less code than I was expecting for certificate auth.

Now, if we have to start providing docs for how to do certificate auth *right*, I think the amount of changes grows substantially, but I don't think it's really worth it at this point.

TomSweeneyRedHat commented:
Sending the directory location of the private key out over the wire doesn't give me the warm fuzzies.  Is it something that other software does?

jwhonce commented:
@Luap99 I seem to remember the idea behind using the ssh executable as well as the golang libraries was to push this kind of handshake into that code vs. podman. But @cdoern did that work, so I don't know where it stands.

Luap99 commented:
> > Sending the directory location of the private key out over the wire doesn't give me the warm fuzzies. Is it something that other software does?
> 
> That certainly wasn't my intent, but I'm not sure what you're referring to. Is that a consequence of adding it to the conf file that I'm unaware of?

It is not being send anywhere, both the client and server read the files locally and then use the certificates to perform a normal TLS handshake via the go std lib AFAICT.
@TomSweeneyRedHat What are you referring to by sending the private key over the write?




> @Luap99 I seem to remember the idea behind using the ssh executable as well as the golang libraries was to push this kind of handshake into that code vs. podman. But @cdoern did that work, so I don't know where it stands.

@jwhonce  I am not sure what are you referring to? The question for me was if there was a specific reason why TLS support was not done before. If not then I think we should accept this PR

Luap99 commented on pkg/util/tls.go:
```diff
@@ -0,0 +1,29 @@
+	"os"
+)
+
+func ReadCertBundle(path string) (*x509.CertPool, error) {
```
can you mope this in a new separate package, maybe call it `cert` and then name the function ReadBundle()

We should really move away from dumping everything into util packages as this causes a lot of unwanted side effects via other imports that really should not have to be imported on the remote client, i.e. https://github.com/containers/podman/issues/23818. Fixing that will be quite some work but in the meantime we should not add new things there that are needed by the remote client.

meln5674 replied:
Done

Luap99 commented on pkg/bindings/connection.go:
```diff
@@ -298,11 +316,31 @@ func tcpClient(_url *url.URL) (Connection, error) {
+		DialContext:        dialContext,
+		DisableCompression: true,
+	}
+	if len(tlsCAFile) != 0 || len(tlsCertFile) != 0 || len(tlsKeyFile) != 0 {
+		logrus.Debugf("using TLS cert=%s key=%s ca=%s", tlsCertFile, tlsKeyFile, tlsCAFile)
+		transport.TLSClientConfig = &tls.Config{}
+		connection.tls = true
+	}
+	if len(tlsCAFile) != 0 {
+		pool, err := util.ReadCertBundle(tlsCAFile)
+		if err != nil {
+			return connection, fmt.Errorf("unable to read CA bundle: %w", err)
+		}
+		transport.TLSClientConfig.RootCAs = pool
+	}
+	if len(tlsCertFile) != 0 && len(tlsKeyFile) != 0 {
```
The condition here do not seem to line up perfectly.

If only `tlsCertFile` is set then the first condition matches and says tls is used but then we never add the the key on the last one. That case should return a hard error I would assume.

meln5674 replied:
My intent was to have that check performed in the `cmd/podman/...` packages, but on reflection it makes sense to also perform the same check here, as that wouldn't account for the env vars. Done.

Luap99 commented on cmd/podman/system/connection/add.go:
```diff
@@ -141,11 +157,29 @@ func add(cmd *cobra.Command, args []string) error {
 
 	switch uri.Scheme {
 	case "ssh":
+		if cmd.Flags().Changed("tls-cert") {
+			return errors.New("--tls-cert option not supported for ssh scheme")
+		}
+		if cmd.Flags().Changed("tls-key") {
+			return errors.New("--tls-key option not supported for ssh scheme")
+		}
+		if cmd.Flags().Changed("tls-ca") {
+			return errors.New("--tls-ca option not supported for ssh scheme")
+		}
 		return ssh.Create(entities, sshMode)
 	case "unix":
 		if cmd.Flags().Changed("identity") {
 			return errors.New("--identity option not supported for unix scheme")
 		}
+		if cmd.Flags().Changed("tls-cert") {
+			return errors.New("--tls-cert option not supported for unix scheme")
+		}
+		if cmd.Flags().Changed("tls-key") {
+			return errors.New("--tls-key option not supported for unix scheme")
+		}
+		if cmd.Flags().Changed("tls-ca") {
+			return errors.New("--tls-ca option not supported for unix scheme")
+		}
```
This duplicates the same conditions. Move them outside the switch case and match `if scheme != tcp` so you do not duplicate that

meln5674 replied:
Done

Luap99 reviewed (commented):
Also this will need some e2e or system tests where we spawn a server with TLS and the connect with the remove client.

meln5674 commented:
> Also this will need some e2e or system tests where we spawn a server with TLS and the connect with the remove client.

Of course. I was hoping to have my environment capable of passing tests on main before I went on vacation last week, but that didn't end up happening. After throwing up my hands and just making a fresh CentOS VM, I have that working now. 

I'd like to use any existing tests for SSH/TCP remotes as a basis for my tests, as well as ensure my tests have parity with them, but for the life of me, I cannot seem to find any such tests. Would you mind pointing me to where they are?

TomSweeneyRedHat commented:
@Luap99 sorry, just back to this.  This line: https://github.com/containers/podman/pull/24601/files#diff-cf5cc76e70a369c5221f08f29ec0452710af1f4ecd423c162d5b669688416caeR525  seemed to indicate that you were looking for the directory where the private key resigned.  Unless I misinterpreted, that doesn't feel right.

meln5674 commented:
> @Luap99 sorry, just back to this. This line: https://github.com/containers/podman/pull/24601/files#diff-cf5cc76e70a369c5221f08f29ec0452710af1f4ecd423c162d5b669688416caeR525 seemed to indicate that you were looking for the directory where the private key resigned. Unless I misinterpreted, that doesn't feel right.

Yes, the purpose of this line is to create a command line argument so that the user can provide the path to their TLS client private key, which is necessary to perform authentication. Unless I too have misinterpreted something else in the codebase, neither that path, nor especially not the data in that file, ever goes over the wire. This is the analog of the `--identity` flag for SSH authentication.

meln5674 commented:
> Also this will need some e2e or system tests where we spawn a server with TLS and the connect with the remove client.

Apologies for the delay on this, a number of other items were pushed onto my priority stack, but I've been able to resume work on it.

As mentioned, there weren't any substantial e2e tests for the TCP API, and the libpod remote test suite was entirely using the unix socket, so I have refactored it to allow running it for unix, plain TCP, TLS, and mTLS. This might seem excessive, but it was only by running this complete suite did I find additional issues I had to fix. Suggestions are welcome for how to achieve similar coverage with fewer tests.

Speaking of said issues, it would seem that the container attach client has a subtle race condition in it, as well as a memory leak.

By overwriting the http transport and capturing the net.Conn as a local variable, if the same Connection is used concurrently, there is a possibility that the first net.Conn will be overwritten on the first attach call's local variable, resulting in the two calls proxying each-other's streams, or just outright corrupting them.

As well, because the original transport is never restored, if the same client is used to attach to multiple containers, the closure is never freed, meaning the dial function will call an ever-increasing stack of closures. Fixing this issue felt outside the scope of this PR, but I have added a note warning of it to whoever feels like addressing it in the future.

I would suggest adding some sort of `DoRequestWithConn` method which handles using the new transport for that request alone and manually performs the dialing, writing the request, and reading the response, and validating the upgrade.

Finally, there were a number of other e2e tests that were failing in my fresh centos 9 environment, even on main, so I have simply ignored them for the time being.

Luap99 reviewed (commented):
just a few comments, I need more time to look at the test changes. But we definitely cannot just explode the matrix and run 4 times the amount.

meln5674 commented on Makefile:
```diff
@@ -682,7 +682,10 @@ ginkgo:
 .PHONY: ginkgo-remote
 ginkgo-remote:
-	$(MAKE) ginkgo-run TAGS="$(REMOTETAGS) remote_testing"
+	$(MAKE) ginkgo-run TAGS="$(REMOTETAGS) remote_testing remote_unix_testing"
+	$(MAKE) ginkgo-run TAGS="$(REMOTETAGS) remote_testing remote_tcp_testing"
+	$(MAKE) ginkgo-run TAGS="$(REMOTETAGS) remote_testing remote_tls_testing"
+	$(MAKE) ginkgo-run TAGS="$(REMOTETAGS) remote_testing remote_mtls_testing"
```
I agree, this was a "nuclear option" in order to get tests to run with the least amount of change to the tests themselves, and to make sure that I wasn't missing anything.

One thought I had in the interim was that ginkgo allows tagging tests and sets of tests, and selecting a subset of them on the command line. Using this, it would be possible to run a full remote tests via unix sockets, and then a subset of all tests against tcp, tls, and mtls. It would then also be possible to add a separate target to run all tests in all remotes, but wouldn't be run in CI, only on-demand in development environments.

The main thing I would need for this is guidance from the podman core team on which tests they believed were critical to be tested over every possible remote, and which are "good enough" to only be tested over unix. For example, the attach endpoints would definitely need to be tested on all different remotes, as that's how I discovered the additional fixes I had to make.

meln5674 replied:
I haven't heard back, so I've gone ahead with this strategy with a small selection of, admittedly arbitrary, tests to pare down to when testing TCP, TLS, and mTLS, based mostly on gut feeling of "things that might go wrong when using a different transport".

ashley-cui replied:
@Luap99 WDYT of the current labeling tests solution vs running past-fedora tests using tls? The labels are a clever spot checking solution, but I think editing the cirrus matrix for split testing feels more maintainable?

meln5674 commented on cmd/podman/system/connection/list.go:
```diff
@@ -118,19 +118,22 @@ func inspect(cmd *cobra.Command, args []string) error {
 	} else {
 		rpt, err = rpt.Parse(report.OriginPodman,
-			"{{range .}}{{.Name}}\t{{.URI}}\t{{.Identity}}\t{{.Default}}\t{{.ReadWrite}}\n{{end -}}")
+			"{{range .}}{{.Name}}\t{{.URI}}\t{{.Identity}}\t{{.TLSCAFile}}\t{{.TLSCertFile}}\t{{.TLSKeyFile}}\t{{.Default}}\t{{.ReadWrite}}\n{{end -}}")
```
I *would* agree, and while I'm not sure if podman or the the umbrella containers project has strict guidelines on the topic, I've always considered "human readable" data to be excluded from breaking changes. I would no sooner worry about breaking scripts that scrape this output than scripts that break if a new log message was added.

meln5674 replied:
After thinking about this for a bit, if breaking changes are that much of a concern, I have hidden the new outputs behind a flag. My thought is that this flag can be removed and this output made the default at the next major version bump when breaking changes are permissible.

ashley-cui replied:
Looks like this is breaking some system tests ATM.

ashley-cui replied:
I think this still counts as a breaking change ATM. I think the best thing to do is just to drop them, or to move them to the end, after readwrite. I prefer to drop them, and just document them as format options, so that if a user wants to see them, can can use `--format ` to display them

meln5674 commented on test/e2e/system_connection_test.go:
```diff
@@ -37,16 +38,20 @@ func setupConnectionsConf() {
 
 	Context("without running API service", func() {
 		It("add ssh://", func() {
-			cmd := []string{"system", "connection", "add",
```
I believe all of the errant auto-formats have been reverted. Please let me know if I have missed any.

meln5674 commented:
Will do. I'll need to spend a bit of time resurrecting the dev VM I had for this today to make sure tests still all pass locally after rebasing.

meln5674 commented:
Less progress than I would have liked today, had to spend time putting my homelab virtualization cluster back together after it had been scavenged for parts over the intervening months. Rebased both podman and common, and added the go.mod replacement and re-vendored to get CI going.

Unit tests are passing locally. Remote E2E tests for the ssh/unix transport are flaking, really badly, as in, "can't get a successful run" kind of badly. This is odd, since that wasn't the subject of any of my changes, and the TCP/TLS/mTLS E2E tests work fine, which are all the same tests, just over a different transport. Errors are consistently about permissions issue of a `layer.lock` file. Not sure if this an issue resulting from changes on main, or something else I have overlooked. 

There seems to have been a test added in the last 6 months that doesn't make sense within the idea of multiple remote transports. I have disabled it temporarily in order to get CI going.

Tomorrow, or failing that, over the weekend, I hope to integrate that new test with the tests for multiple transports, as well as add a few last tests in order to confirm that every combination of getting the tls material from connection/flags/env/none works as expected, and finally address any CI findings.

mheon commented:
Bloat approved tag added, CI restarted

meln5674 commented:
Sorry for the radio silence on this. In addition to some 9-to-5 stuff preempting this, I was seeing a failure on [this system test](https://github.com/containers/podman/blob/main/test/system/090-events.bats#L32) and, despite it being unrelated to this PR, was convinced it was an issue on my end. Not sure why it didn't occur to me to check this sooner, but its also failing on main (c73712b66337ff0d954d4acd196da7bc42da4a65) on a VM from a [fedora cloud image](https://fedoraproject.org/cloud/download) from a week ago.

For the time being, I rebased to fix the conflicts and fixed the other failing system test you noted above to see if it will also fail in CI.

ashley-cui commented on pkg/bindings/connection.go:
```diff
@@ -93,14 +96,34 @@ func NewConnection(ctx context.Context, uri string) (context.Context, error) {
+func NewConnectionWithOptions(ctx context.Context, opts Options) (context.Context, error) {
+	var err error
+
+	uri := orEnv(opts.URI, "CONTAINER_HOST")
+	identity := orEnv(opts.Identity, "CONTAINER_SSHKEY")
+	tlsCertFile := orEnv(opts.TLSCertFile, "CONTAINER_TLS_CERT")
+	tlsKeyFile := orEnv(opts.TLSKeyFile, "CONTAINER_TLS_KEY")
+	tlsCAFile := orEnv(opts.TLSCAFile, "CONTAINER_TLS_CA")
```
I think the env var lookup should be done in containers/common. There, we can set the values to be read first from the env var, then from the containers.conf value.

ashley-cui replied:
Still applied. Should be plumbed through the CLI through common.

ashley-cui commented on cmd/podman/system/connection/list.go:
```diff
@@ -118,19 +118,22 @@ func inspect(cmd *cobra.Command, args []string) error {
-			"ReadWrite": "ReadWrite",
+			"Default":     "Default",
+			"Identity":    "Identity",
+			"TLSCAFile":   "TLSCAFile",
+			"TLSCertFile": "TLSCertFile",
+			"TLSKeyFile":  "TLSKeyFile",
```
ditto

meln5674 replied:
Also, I'm not certain I understand how the second is a breaking change, those are just the list of fields *available* for use, not *required* to be used, in a format string. If those are not there, a user wouldn't even be able to `--format` to see them.

ashley-cui reviewed (commented):
@meln5674 First of all, thanks for all your work! We truly appreciate community contributions, and apologize that this has slipped through the cracks for so long


> It doesn't look like the failure I'm seeing locally is reflected in CI, thankfully, but the two new failures are both external:

Apologies for running into flakes, our CI is a constant WIP. Feel free to open any issues to help us keep track of what's going wrong

> 
> The last thing I'd like to do is add the system tests that verify the combinations of flags, connections, and env vars for TLS options, but I'm unsure of you would like me go about doing that, as it touches on the same issues as the integration tests. 

Unfortunately, I think the tests in their current state needs to be reworked. Labeling certain tests to be run several times is unmaintainable for us, and makes it difficult for contributors to know why a test is failing. My suggestion is to do this as a system test rather than e2e. Looking at the [cors tests](https://github.com/containers/podman/blob/main/test/system/271-tcp-cors-server.bats) for inspiration might be a great starting point. I think its sufficient just to test that mTLS/TLS works, and fails when an incorrect cert is given or a key is missing.

As for the flags/connections/env vars I think the best thing to do is: 
- Test the flags in podman, using the system tests strategy mentioned above
- Test that the system connections are properly added and listed in the [e2e tests (https://github.com/containers/podman/blob/main/test/e2e/system_connection_test.go) If you would like to test functionality, I think the best place for that is in the system tests mentioned above.
- Test the env vars in c/common as mentioned in the review

Lastly, I know this is still a WIP so it's fine for now to make it easier for revisions, but once it's no longer a draft and c/common has merged, squash your commits, but leave the c/common vendoring as a separate commit. 

Thanks again for the tremendous amount of work!

meln5674 commented:
@ashley-cui 
> I think its sufficient just to test that mTLS/TLS works, and fails when an incorrect cert is given or a key is missing.

When I started working on this, I would have agreed with you, but there is an unfortunate wrinkle that necessitates at least *some* granular testing of tls/mtls.

If the only change were to how the connection was initiated, then a simple smoke test like you're describing would be adequate. However, [these](https://github.com/containers/podman/pull/24601/files/c73712b66337ff0d954d4acd196da7bc42da4a65#diff-02a7031ca65fa34d69f3de533b25111aa07a03242c86c4660dbd35ec69d1cd2fR136) two [sections](https://github.com/containers/podman/pull/24601/files/c73712b66337ff0d954d4acd196da7bc42da4a65#diff-02a7031ca65fa34d69f3de533b25111aa07a03242c86c4660dbd35ec69d1cd2fR508) had to be added in order to make the "streaming" functionality work. This is a result of using a non-standard approach to perform the HTTP connection upgrade to the "raw" protocol originally invented by docker (similar concept to websockets for anyone whom that's a more familiar topic).

What this means is, there could be functionality that is silently broken that would not be surfaced until someone attempted to use *every* piece of functionality over a TLS connection. If new functionality were to be added in the future that also leveraged this connection upgrade, that functionality could be released in a broken state for TLS connection, without noticing, as it would never be executed over TLS during CI.

I agree that the method I have so far is sub-optimal, and needs to be reworked, but I strongly believe that the other connection methods need to have at least *some* robust testing in CI beyond '`podman ps` works, ship it'.

meln5674 commented:
Having slept on the problem, I want to take a step back and ensure there aren't any conflicting unspoken assumptions. To be explicit, my goals with the tests the way I have/was planning on doing them is to prove the following:

1. Existing unix/ssh remote functionality works as it did before
2. Desired properties of (m)TLS hold (client rejects untrusted server, server rejects untrusted server, neither attempt to run with missing/malformed material)
3. Existing functionality (running containers, streaming logs, etc) works over the new transports

The way I intended to prove those were:

1. Existing tests execute without modification
2. Tests which contrive scenarios where those rejections occurs (bogus cert, garbage data in cert file, outright missing, etc)
3. Run a small selection of existing test cases against the new transports (e.g. run a container, run with attached streams, stream logs, pull an image, etc)

You've made mention of the tests in point 2, so it appears we agree on that. Can you please clarify if the disagreement for point 3 is purely in implementation, or do you disagree that it needs to be proven at all?

ashley-cui commented:
@meln5674 Thanks for the clarification! 3 is mostly on implementation. The labels solution is just too difficult to maintain, and also makes it difficult to understand why that test is failing and if it's run with a different transport. Beyond that, I think it's a good idea to test some core podman functionality with TLS, something like testing image pull, running a container, doing a build, and a few more test cases where you see fit might be okay. Running the entire suite with TLS seems to be too much to me, I think it's enough to trust our split testing there. If bugs arise, we can always fix them, and then add a specific test case for that bug as they are fixed.

meln5674 commented on test/e2e/image_scp_test.go:
```diff
@@ -23,6 +23,7 @@ var _ = Describe("podman image scp", func() {
 	})
 
 	It("podman image scp with proper connection", func() {
+		Skip("TODO: This test doesn't make sense in the context of multiple remote transports")
```
I think I added that in an earlier commit and forgot about it. Do you think it'd make sense to have a set of transport-specific integration tests akin to the transport-specific system tests that this'd be a part of?

ashley-cui replied:
I think I'd prefer to just write a system test. We've been wanting to de-dup between our system and e2e anyway.

meln5674 replied:
Done

meln5674 commented on Makefile:
```diff
@@ -695,44 +700,74 @@ localsystem:
+	REMOTESYSTEM_TLS_CLIENT_CRT \
+	REMOTESYSTEM_TLS_CLIENT_KEY
+
+$(REMOTESYSTEM_TLS_CA_CRT) $(REMOTESYSTEM_TLS_CA_KEY) &:
```
I can move it to an all-in-one script that regenerates all of the certs/keys at once, and have a single target instead of separate ones if that'd be more maintainable. I think having at least the option to run the full suite against a non-unix remote is valuable, as it already found a few places were the system tests were making implicit assumptions which I was able to make explicit.

meln5674 replied:
Done

meln5674 commented:
@ashley-cui 
I have added the tests I described to verify the different remotes and their options, and I'm glad I did, as they uncovered a few more bugs of mine I hadn't noticed, including a panic. I have also re-done the test certificate generation as a single target calling a script. I am tracking the two outstanding requests regarding moving the env vars to common and the --format=tls option, but I'd like a thumbs-up on the testing situation before I move onto that. I also noticed that my autoformatted somehow reactivated, so I've given up, and I will just revert the autoformatting at the end when I squash.

meln5674 commented on test/system/273-remote-spot-check.bats:
```diff
@@ -0,0 +1,127 @@
+@test "unix remote" {
+  URL=unix:$PODMAN_TMPDIR/myunix.sock
+
+  systemd-run --unit=$SERVICE_NAME ${PODMAN%%-remote*} system service $URL --time=0
```
I thought so too, but it looks like other similar tests runs for both remote and local suites. I suspect the intent was to test both the full podman and standalone podman-remote binary. I'm fine with either way.

meln5674 commented on test/system/273-remote-spot-check.bats:
```diff
@@ -0,0 +1,127 @@
+  wait_for_file $PODMAN_TMPDIR/myunix.sock
+
+  # Variable works
+  CONTAINER_HOST=$URL run_podman info --format '{{.Host.RemoteSocket.Path}}'
```
I'm not certain I follow. I agree with moving the code that reads the environment variables to common and testing that they are read correctly, but that wouldn't test that podman itself is actually using them. Am I misunderstanding and you're actually suggesting moving the entire connection initiation code to common and testing it that way?

meln5674 replied:
I originally added the variables to mirror how existing options were handled, i.e CONTAINER_HOST for --host, CONTAINER_IDENTITY for --identity. As well, for those coming from docker, there the DOCKER_CERT_PATH variable, albeit, which works differently. It definitely adds implementation complexity, but I find its a very common pattern that people come to expect, that is, having a scope-narrowing precedence chain of
1. cli flag (scoped to a single command or process) 
2. env var (scoped to a single terminal session via exports) 
3. config file entry (global scope)
I know I often make use of this pattern in a number of adjacent tools, e.g. kubectl, helm, containerd, etc, though I'd be interested to hear other perspectives.

meln5674 replied:
I am currently re-working the system tests to be able to handle the different remotes without have environment variables as an option, and I have to say, the additional complexity of passing around flags is an order of magnitude greater than it was when using envs. I can only imagine attempting to use it "for real" will be even worse. If this is anything to go by, using remotes in any sort of automation will be serious headache.

On the off chance the decision to remove the TLS envs were based on that complexity, I would strongly suggest  re-evaluating this if that aspect of the UX is a priority.

meln5674 commented on test/system/251-system-service.bats:
```diff
@@ -31,8 +31,11 @@ function teardown() {
-    run_podman --host $URL info --format '{{.Host.RemoteSocket.Path}}'
-    is "$output" "$URL" "RemoteSocket.Path using unix:"
+    (
+      unset CONTAINER_HOST CONTAINER_TLS_{CA,CERT,KEY}
```
We would, at minimum, need to unset the TLS variables for the tcp test, as those are set in the makefile when selecting which transport to run the tests against, otherwise the tcp test would attempt to connect to an (m)tls server, which would fail. The test itself is assuming that they unset, and this makes that assumption explicit. I could selectively unset specific variables for each test, but I felt that consistency was the better approach.

ashley-cui replied:
Assuming we're going with the removal of the env vars, this shouldn't be an issue. We would then not need to unset the tls variable,s and then just ensure the flag is set for the tls tests.

ashley-cui commented:
FYI we migrated c/common into a monorepo today - you need to move your c/common PR to c/container-libs, apologies for the inconvenience.

meln5674 commented on pkg/bindings/containers/attach.go:
```diff
@@ -130,7 +133,28 @@ func Attach(ctx context.Context, nameOrID string, stdin io.Reader, stdout io.Wri
+			if err != nil {
+				return nil, err
+			}
+			cfg := tlsConfig.Clone()
```
This code is not called for non-TLS connections. However, the value could have been mutated between construction and this call, so I will defensively add a nil check to allocate a new blank config instead of copying a nil one.

meln5674 commented on pkg/bindings/containers/attach.go:
```diff
@@ -115,9 +116,11 @@ func Attach(ctx context.Context, nameOrID string, stdin io.Reader, stdout io.Wri
 	dialContext := conn.Client.Transport.(*http.Transport).DialContext
+	tlsConfig := conn.Client.Transport.(*http.Transport).TLSClientConfig
 	t := &http.Transport{
 		DialContext: func(ctx context.Context, network, address string) (net.Conn, error) {
```
What are you referring to by "this"? The `DialContext` field? `DialContext` is still called even if `DialTLSContext` is set, *if the connection is non-TLS*, so we cannot remove it, unless your suggestion is to inspect `conn.URL.Scheme` and set one or the other conditionally, which feels like extra complexity for no benefit.

mtrmac commented on pkg/bindings/connection.go:
```diff
@@ -127,13 +150,13 @@ func NewConnectionWithIdentity(ctx context.Context, uri string, identity string,
 		connection = conn
 	default:
-		return nil, fmt.Errorf("unable to create connection. %q is not a supported schema", _url.Scheme)
+		return nil, fmt.Errorf("unable to create connection. %q is not a supported schema. %#v %s %s", _url.Scheme, opts, uri, _url.String())
```
Sure, that works.

mtrmac commented:
> It looks like go.mod still points to common instead of container-libs, is there an expected timeline for the switch?

That happened just now, in https://github.com/containers/podman/pull/26858 .

ashley-cui commented:
> Testing-wise: I wonder if it wouldn't be easiest to have fedora N be non-TLS, and fedora N-1 be TLS, for the E2E tests (system tests look pretty reasonable, running the TLS bits separate seems fine there). We've done a split across OSes before (OCI runtime testing, for example) and it has worked well - Fedora N vs N-1 differences are generally small enough that we can get away with it.

Do you think that it's necessary to run our entire test suite with tls? I honestly think its suffiencient to just run a few system tests under tls. Maybe if we do decide to do split testing, it can be addressed in a separate PR.

ashley-cui commented:
@meln5674  Friendly ping, still working on this?

meln5674 commented:
@ashley-cui I was waiting to hear back re: the env vars conversation above, and then some personal stuff came up. I should be able to resume within the next few days. I have mtrmac's feedback implemented, however, I wanted to get a chance to discuss the env vars again before I committed and rebased to the container-libs switch.

I still strongly believe that envs should be an option. As I mentioned, many tools in the space do this, such as the KUBECONFIG variable in kubectl and helm, the DOCKER_CERT_PATH for docker, CONTAINERD_ADDRESS for containerd, and so on. I'm still not sure I understand the motivation behind removing them is, and if it comes down to shaving a few seconds off the system tests, that feels like a poor tradeoff. Given that I'm already running into frustrations doing simple things as part of the system tests, I imagine anyone trying to use it for real would find the flags as the only override options to be an obstacle.

ashley-cui commented:
@meln5674 The reason behind removing the environment variables is that we cannot support env vars for every option that exists. Persistent configurations should be set in containers.conf, and one-off configurations should be set using a flag. Therefore, both the use cases for env vars are covered, so there's no reason to support an env var as well. 

What issues are you running into when removing the env var? Please feel free to push to another branch and link it, maybe I can help out there. 

As an aside, we're approaching a deadline soon of when we want this feature in. I'd strongly prefer only working on the system tests, and the integration and makefile target can be worked on in a followup PR, as it looks like there are still design discussions around that as well.

meln5674 commented:
Unfortunately, its the system tests that are the issue, integration is running fine, and it really comes down to the decision to write the system tests in bash. Before, I could export the appropriate envs from the makefile helper script, and for tests that needed to use their own, it would unset them at the top of that test. I can no longer do that, so I have to invent new variables, and build the argument lists in a help function, or use global mutable arrays, except some tests don't use the helper functions, or use them in a way that conflicts with how other tests expect to use them, or some tests use subshells which don't get arrays, so I have to go through and fix or fudge a decent chunk of the other tests that have nothing to do with this feature. Its not that its a difficult engineering challenge, its just tedious due to how its written.

I should have time later today to pick this up and report back with specifics. Can you elaborate on the deadline?

mtrmac commented:
> I wouldn't want them documented, but maybe testing-only ways of setting up TLS via environment is not the worst idea?

A somewhat weak opinion: A “testing-only undocumented” way to affect security settings instinctively makes me worried about unexpected / hard-to-audit ways to affect configuration. (It would be plausible to have testing-only ways to affect variables _that are only possible in testing builds_, but then we would not be testing exactly the production code.)

meln5674 commented:
Eugh. I've always maintained the position that if you have to put undocumented things in production code to facilitate testing, you don't have a testable architecture.

mtrmac commented on cmd/podman/system/connection/list.go:
```diff
@@ -114,23 +120,31 @@ func inspect(cmd *cobra.Command, args []string) error {
-		rpt, err = rpt.Parse(report.OriginUser, format)
-	} else {
+	switch format {
+	case "tls":
```
There’s some precedent to special-casing `--format=…` values (in particular we accept --format json` in places), and to an extent I appreciate the ergonomic, but, on the _principle_ of the thing, I personally dislike values that have two different semantics. If `--format hello` outputs literal "hello", it’s inconsistent that `--format tls` does not output literal "tls".

I’m not authoritative for Podman, and note that this _was_ pre-approved in https://github.com/containers/podman/pull/24601#discussion_r2270716090 .

mtrmac commented on test/e2e/common_test.go:
```diff
@@ -278,8 +284,18 @@ func getPodmanBinary(cwd string) string {
+	PodmanTestCreateUtilTargetTCP   = "tcp"
+	PodmanTestCreateUtilTargetTLS   = "tls"
+	PodmanTestCreateUtilTargetMTLS  = "mtls"
+)
```
If the intent is to create typed enum-like constants, this does not do that — every const line needs to explicitly include the `…UtilTarget` type.

mtrmac commented on test/e2e/system_connection_test.go:
```diff
@@ -37,16 +38,21 @@ func setupConnectionsConf() {
+	systemConnectionListCmd                = []string{"system", "connection", "ls", "--format", "{{.Name}} {{.URI}} {{.Identity}} {{.Default}} {{.ReadWrite}}"}
+	systemConnectionListTLSSpecialValueCmd = []string{"system", "connection", "ls", "--format", "tls"}
+	systemConnectionListTLSCmd             = []string{"system", "connection", "ls", "--format", "{{.Name}} {{.URI}} {{.TLSCAFile}} {{.Default}} {{.ReadWrite}}"}
+	systemConnectionListmTLSCmd            = []string{"system", "connection", "ls", "--format", "{{.Name}} {{.URI}} {{.TLSCAFile}} {{.TLSCertFile}} {{.TLSKeyFile}} {{.Default}} {{.ReadWrite}}"}
```
(I think moving the commands away from the core of the test code only impedes readability compared to including the in-line in the test, but I’ll fully defer to Podman maintainers on this.)

Luap99 commented on Makefile:
```diff
@@ -702,51 +708,52 @@ localmachine:
 	@echo /define.gitCommit=$(GIT_COMMIT)
 	$(MAKE) ginkgo-run GINKGO_PARALLEL=n TAGS="$(REMOTETAGS)" GINKGO_FLAKE_ATTEMPTS=0 FOCUS_FILE=$(FOCUS_FILE) GINKGOWHAT=pkg/machine/e2e/.
 
+REMOTESYSTEM_TCP_PORT ?= 8080
+REMOTESYSTEM_TLS_CA_CRT ?= $(CURDIR)/bin/remotesystem.ca.crt.pem
+REMOTESYSTEM_TLS_CA_KEY ?= $(CURDIR)/bin/remotesystem.ca.key.pem
+REMOTESYSTEM_TLS_SERVER_CRT ?= $(CURDIR)/bin/remotesystem.server.crt.pem
+REMOTESYSTEM_TLS_SERVER_KEY ?= $(CURDIR)/bin/remotesystem.server.key.pem
+REMOTESYSTEM_TLS_CLIENT_CRT ?= $(CURDIR)/bin/remotesystem.client.crt.pem
+REMOTESYSTEM_TLS_CLIENT_KEY ?= $(CURDIR)/bin/remotesystem.client.key.pem
+REMOTESYSTEM_TLS_BOGUS_CRT ?= $(CURDIR)/bin/remotesystem.bogus.crt.pem
+REMOTESYSTEM_TLS_BOGUS_KEY ?= $(CURDIR)/bin/remotesystem.bogus.key.pem
+
+export \
+	REMOTESYSTEM_TCP_PORT \
+	REMOTESYSTEM_TLS_CA_KEY \
+	REMOTESYSTEM_TLS_CA_CRT \
+	REMOTESYSTEM_TLS_SERVER_CRT \
+	REMOTESYSTEM_TLS_SERVER_KEY \
+	REMOTESYSTEM_TLS_CLIENT_CRT \
+	REMOTESYSTEM_TLS_CLIENT_KEY \
+	REMOTESYSTEM_TLS_BOGUS_CRT \
+	REMOTESYSTEM_TLS_BOGUS_KEY
+
+$(REMOTESYSTEM_TLS_CA_CRT) $(REMOTESYSTEM_TLS_CA_KEY) \
+$(REMOTESYSTEM_TLS_CLIENT_CRT) $(REMOTESYSTEM_TLS_CLIENT_KEY) \
+$(REMOTESYSTEM_TLS_SERVER_CRT) $(REMOTESYSTEM_TLS_SERVER_KEY) \
+$(REMOTESYSTEM_TLS_BOGUS_CRT) $(REMOTESYSTEM_TLS_BOGUS_KEY) \
+&:
+	source hack/remotesystem.env ; remotesystem-gen-tls
```
IMO this has nothing to do in the Makefile, it bloats the Makefile with a bunch of unrelated options.

More importantly it creates a dependency on the makefile targets to run the bast tests which is simply wrong.

The bats tests must be able to be run on their own via just `bats test/system`, in fact we just ship the test/system directory in the test/system rpm. to run them as part of distro level gating on reverse dependency testing.
There is also the the way to run hack/bats for local dev testing.

So if we need additional setup such as creating temporary certs it must always be encapsulated in the bats setup themselves.

meln5674 replied:
I can move the actual cert generation to the bats setup helpers, but these variables must be known to the makefile, as the makefile previously and still does, now through hack/remotesystem.env, perform a check to ensure there is no podman service already running as well as to wait for the newly launched service to start.

Luap99 commented on Makefile:
```diff
@@ -702,51 +708,52 @@ localmachine:
-		echo "Skipping $@: 'timeout -v' unavailable'";\
-	fi;\
-	exit $$rc
+	source hack/remotesystem.env ; \
+	set -x ; \
+	remotesystem-ensure-timeout-cmd $@ || exit ; \
+	remotesystem-podman-service $(REMOTESYSTEM_TRANSPORT) $(PODMAN_SERVER_LOG) && \
+	remotesystem-wait-podman-service $(REMOTESYSTEM_TRANSPORT) && \
+	remotesystem-bats $(REMOTESYSTEM_TRANSPORT) $(CURDIR)/bin/podman-remote
```
I would just have the script do everything directly instead of spiting this awkwardly with source here into all these functions.

Also I think it is safe to drop the `timeout` check from it. It really servers no purpose to skip the tests, we can just let them fail hard in that case.

meln5674 replied:
I intended to reproduce the embedded script as it was, but if that part is no longer necessary, I will remove it. I will bundle these all into a single "remotesystem-run" function.

Luap99 commented on test/system/850-compose.bats:
```diff
@@ -6,8 +6,8 @@
 @test "podman compose - smoke tests" {
-    fake_compose_bin="$PODMAN_TMPDIR/fake_compose"
-    cat >$fake_compose_bin <<EOF
+  fake_compose_bin="$PODMAN_TMPDIR/fake_compose"
+  cat >$fake_compose_bin <<EOF
```
these formatting changes are totally unrelated, please drop them as this bloats the diff for no reason

meln5674 replied:
Reverted

Luap99 commented on test/system/272-system-connection.bats:
```diff
@@ -125,13 +133,139 @@ $c2[ ]\+tcp://localhost:54321[ ]\+true[ ]\+true" \
+    # Add the connection, and run podman info *before* starting the service.
+    # This should fail.
+    run_podman system connection add myconnect tcp://localhost:$_SERVICE_PORT \
+      --tls-ca="${REMOTESYSTEM_TLS_CA_CRT}"
```
see my comments in the Makefile you cannot ever amuse that these carts exists here in the tests, if tests need their own certs they must be generated in the bats setup not externally. Same for all the tests you changed here.

Luap99 replied:
No they do in fact not do that, `bats -f test/system` works because helpers.bash defines it for us
`PODMAN=${PODMAN:-podman}`

Just after this review today I was pinged by some folks who also happen to run these tests directly in all cases the supported invocation is `bats test/system` for them. We should not break them and require the existence of additional files to set the test up.

On fedora you can do `dnf install podman podman-tests` and then `bats /usr/share/podman/test/system` so that is a usecase we must continue to support as this is uses as gating tests, no need to have the podman source git repo at all. So it is critical that these tests do not depend on anything in the repo outside of test/system.

Sure for remote testing extra setup is required outside of the test suite but for most basic gating/reverse dependency testing that is not needed as they only use the local testing.

Luap99 commented on test/e2e/common_test.go:
```diff
@@ -622,7 +784,7 @@ func (p *PodmanTestIntegration) RunTopContainer(name string) *PodmanSessionInteg
 func (p *PodmanTestIntegration) RunTopContainerWithArgs(name string, args []string) *PodmanSessionIntegration {
 	// In proxy environment, some tests need to the --http-proxy=false option (#16684)
-	var podmanArgs = []string{"run", "--http-proxy=false"}
+	podmanArgs := []string{"run", "--http-proxy=false"}
```
here as well please avoid all formatting changes like this. IT makes it harder to review.

meln5674 replied:
Reverted

Luap99 commented on test/e2e/common_test.go:
```diff
@@ -404,15 +426,155 @@ func PodmanTestCreateUtil(tempDir string, remote bool) *PodmanTestIntegration {
+			if err == nil {
+				lockFile.Close()
+				p.RemoteSocketLock = lockPath
+				lis, err := net.Listen("tcp", "127.0.0.1:0")
+				if err == nil {
+					defer lis.Close()
+					p.RemoteSocket = fmt.Sprintf("tcp://%s", lis.Addr())
```
this is not race safe, if we do this for every single test I can guarantee this will flake quite often.
the port is freed before the service can bind it. You would need to pick ports not from the dynamic range and lock them so other tests don't use it in parallel. Or maybe better we make use of the socket activation feature instead. We can pass the fd to the podman service by using the systemd socket activation protocol.

meln5674 commented:
@ashley-cui Is the above sufficient to get the matching container-libs PR merged?

ashley-cui commented:
@meln5674 Yep, sorry, I added another small nit, but other than that, we're waiting for another maintainer to merge the PR, as we require 2 lgtms.

meln5674 commented:
@ashley-cui Understood. I made the nit'ed changes and rebased this branch on them. Once the container-libs PR is merged, I will point back common back to upstream and squash and all changes to a single commit for final review and CI run.

meln5674 commented:
It doesn't look like container-libs has been assigned a new semantic version. I'll go ahead and re-vendor on the commit hash for now, but should it be like that for the final merge?

meln5674 commented:
Not sure if this is a me problem, but github has been stuck on "Checking for the ability to merge automatically..." for the last hour and change, and [hasn't detected](https://github.com/containers/podman/pull/24601/commits) the [new commit](https://github.com/meln5674/podman/commits/feature/remote-tls/) I pushed. Unsure of what I can do about this.

ashley-cui commented:
@meln5674 Strange, feel free to re-push or to close and re-open the PR, see if that kicks the CI.

mheon commented:
Force-push is usually the most reliable way

meln5674 commented:
One more WSL flake, it looks like.

ashley-cui commented:
LGTM @containers/podman-maintainers PTAL and merge

Luap99 reviewed (approved):
/lgtm
